Tracking/TrackLength/src/TrackLengthUtils.cc

```diff
@@ -46,86 +45,85 @@ IMPL::TrackStateImpl TrackLengthUtils::getTrackStateAtHit(MarlinTrk::IMarlinTrac
+    double omega = std::abs( ts1.getOmega() );
     double z1 = ts1.getReferencePoint()[2] + ts1.getZ0();
     double z2 = ts2.getReferencePoint()[2] + ts2.getZ0();
+    double dz = std::abs(z2 - z1);
+    double dPhi = std::abs( ts2.getPhi() - ts1.getPhi() );
+    // We are never sure whether the track indeed curled for more than pi
+    // or it was just a crossing of the singularity point (-pi/+pi) in the phi coordinate system.
+    // We always assume it was the crossing of the (-pi/+pi) and correct for this.
+    // Thus this formula only applicable for small distances between the track states or for the non-curly tracks (dPhi < pi).
+    if (dPhi > M_PI) dPhi = 2*M_PI - dPhi;
+    return std::sqrt(dPhi*dPhi/(omega*omega)+dz*dz);
+};
 
-    // helix length projected on xy
-    double circHelix = std::abs( (z2-z1)/tanL );
-    double circFull = 2*M_PI/std::abs(omega);
 
-    return circHelix/circFull;
-}
+double TrackLengthUtils::getHelixLengthOption2(const EVENT::TrackState& ts1, const EVENT::TrackState& ts2){
+    double omega = std::abs( ts1.getOmega() );
+    double tanL = std::abs( ts1.getTanLambda() );
+    double dPhi = std::abs( ts2.getPhi() - ts1.getPhi() );
+    // We are never sure whether the track indeed curled for more than pi
+    // or it was just a crossing of the singularity point (-pi/+pi) in the phi coordinate system.
+    // We always assume it was the crossing of the (-pi/+pi) and correct for this.
+    // Thus this formula only applicable for small distances between the track states or for the non-curly tracks (dPhi < pi).
+    if (dPhi > M_PI) dPhi = 2*M_PI - dPhi;
+    return dPhi/omega*std::sqrt(1+tanL*tanL);
+};
+
 
 
 std::vector<EVENT::Track*> TrackLengthUtils::getSubTracks(EVENT::Track* track){
     vector<Track*> subTracks;
+    // add track itself, which contains VXD+FTD+SIT+TPC hits of the first curl.
     subTracks.push_back(track);
 
     int nSubTracks = track->getTracks().size();
     if (nSubTracks <= 1) return subTracks;
 
-    int nTPCHits = track->getSubdetectorHitNumbers()[(ILDDetID::TPC)*2-1];
-    int nSubTrack0Hits = track->getTracks()[0]->getTrackerHits().size();
-    int nSubTrack1Hits = track->getTracks()[1]->getTrackerHits().size();
-
-    //OPTIMIZE: this is not reliable, but I don't see any other way at the moment.
-    //Read documentation in the header file for details.
-    int startIdx;
-    if( std::abs(nTPCHits - nSubTrack0Hits) <= 1  ) startIdx = 1;
-    else if ( std::abs(nTPCHits - nSubTrack1Hits) <= 1 ) startIdx = 2;
-    else{
-        //FIXME: This happens very rarily (0.01%) for unknown reasons, so we just, skip adding subTracks...
-        streamlog_out(WARNING)<<"Can't understand which subTrack is responsible for the first TPC hits! Skip adding subTracks."<<std::endl;
-        return subTracks;
+    auto isTPCHit = [](TrackerHit* hit) -> bool {
+        UTIL::BitField64 encoder( UTIL::LCTrackerCellID::encoding_string() ) ;
```

Comment:
Similar comment as already for the TOFUtils. The construction of the `encoder` should probably be outside the lambda, as I am not entirely sure whether the compiler is clever enough to figure out this optimization opportunity.